ER図とREADME

kei-sasaki-rgb-cell opened the pull request:
# What
ER図とREADMEを作成

ER図Gyazoで撮影
![](https://i.gyazo.com/54217024f558ff06b7a196ce065754ef.png)

# Why
データベース設計に必要なため

kei-sasaki-rgb-cell committed:
ER図とREADME

ky3090 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| Colmun          | Type           | Option            |
+| --------------- | -------------- | ----------------- |
+| item_image      | ActiveStorage  | null: false       |
+| item_name       | string         | null: false       |
```
```
| item_name       | string         | null: false       |
```

itemsテーブルのnameですので、item_は削除しましょう。

ky3090 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| category        |                | null: false       |
+| state           |                | null: false       |
+| postage         |                | null: false       |
+| shipping_date   |                | null: false       |
```
```
| category        |                | null: false       |
| state           |                | null: false       |
| postage         |                | null: false       |
| shipping_date   |                | null: false       |
```

商品の状態, 配送料の負担, 発送元の地域,  発送までの日数, カテゴリーのカラムで不足しているカラムはございませんでしょうか？
再度見直しましょう。

また、商品の状態, 配送料の負担, 発送元の地域,  発送までの日数, カテゴリーのカラムは特別なことがない限り不変な値であることから、今回はactive hashで実装します。
なので、カラムの型をinteger型にしましょう。
そして、カラム名の末尾に_idを記述しましょう。

ky3090 commented on README.md:
```diff
@@ -1,24 +1,73 @@
-* Deployment instructions
+| Colmun          | Type       | Option            |
+| --------------- | -----------| ----------------- |
+| postcode        | integer    | null: false       |
```
```
| postcode        | integer    | null: false       |
```

郵便番号をinteger型にすると先頭の0を補完しないため、string型にしましょう。

ky3090 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| Colmun          | Type       | Option            |
+| --------------- | -----------| ----------------- |
+| postcode        | integer    | null: false       |
+| prefecture_id   |            | null: false       |
```
```
| prefecture_id   |            | null: false       |
```

こちらの都道府県も同様に不変な値ですので、今回はactive hashで実装します。
なのでカラムの型をinteger型にしましょう。

また、itemsテーブルの発送元の地域とこちらの都道府県はどちらも都道府県のデータを使用する関係上
カラム名を統一することで実装がしやすくなります。

ky3090 commented on README.md:
```diff
@@ -1,24 +1,73 @@
+| block           | string     | null: false       |
+| building        | string     |                   |
+| phone_number    | string     | null: false       |
+| user            | references | foreign_key: true |
```
```
| Colmun          | Type       | Option            |
| --------------- | -----------| ----------------- |
| postcode        | integer    | null: false       |
| prefecture_id   |            | null: false       |
| city            | string     | null: false       |
| block           | string     | null: false       |
| building        | string     |                   |
| phone_number    | string     | null: false       |
| user            | references | foreign_key: true |
```

今回住所情報のテーブルは1つで十分です。
なので、上記のカラムは削除しましょう。

また、今回どのユーザーがどの商品を購入したのかを管理するテーブルが必要ですので
そちらを記述しましょう。

ky3090 reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、確認お願い致します！　

kei-sasaki-rgb-cell committed:
ER図、READMEの修正

・itemsテーブルの「item_image」と「item_name」の「item_」の部分を削除

・itemsテーブルのprefecture_id(都道府県カラム)の追加、またactive hashを使う所にはinteger型を記入し、カラムの末尾に_idをつけた。

・shipsテーブルのpostcodeの型をinteger型からstring型に変更。

・prefecture_idの型をinteger型に変更。

kei-sasaki-rgb-cell commented:
修正しました。

![](https://i.gyazo.com/a1ef9f56fe0201968abeb45497a3ee24.png)

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| ----------------- | ------ | ------------------------- |
+| nickname          | string | null: false               |
+| email             | string | null: false, unique: true |
+| password          | string | null: false               |
```
passwordカラムですがdeviseがデフォルトで用意しているencrypted_passwordを使用しましょう！
こちらのカラムにpasswordが暗号化されて保存されます。

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| last_name_kanji   | string | null: false               |
+| first_name_kanji  | string | null: false               |
+| last_name_hira    | string | null: false               |
+| first_name_hira   | string | null: false               |
```
名前と苗字は漢字の場合でも平仮名の場合でも同じカラムに保存されます。
ですのでlast_name、first_nameにまとめた方が良いかと思われます。

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| first_name_hira   | string | null: false               |
+| last_name_kana    | string | null: false               |
+| first_name_kana   | string | null: false               |
+| birthday          | string | null: false               |
```
誕生日の「年、月、日」はdate型を使用すると一つにまとめることができます。
カラムの型をdate型に変更しましょう！

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,70 @@
-* Database creation
+| Colmun           | Type           | Option            |
+| ---------------- | -------------- | ----------------- |
+| image            | ActiveStorage  | null: false       |
```
imageカラムですが、今回は「active_storage」を使用することでimage用の別テーブルが作成され、そちらに保存・管理されます。
ですのでimageカラムは削除しましょう！

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| postage_id       | integer        | null: false       |
+| prefecture_id    | integer        | null: false       |
+| shipping_date_id | integer        | null: false       |
+| price_id         | integer        | null: false       |
```
priceはidで保存されないため「_id」は削除しましょう。

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,70 @@
+
+- belongs_to :user
+- belongs_to :item
+- has_many :ships
```
一つの購入履歴には一つの配送先としたいため「has_one :ship」としましょう！

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,70 @@
+| block           | string     | null: false       |
+| building        | string     |                   |
+| phone_number    | string     | null: false       |
+| user            | references | foreign_key: true |
```
ユーザーテーブルは配送先テーブルと紐付ける必要がないため、
userの代わりにbuyerの外部キーカラムを用意しましょう！

Saka-toku reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
ER図も確認しやすく素晴らしいです！
追加でコメントいたしましたので、ご確認お願いいたします！

kei-sasaki-rgb-cell committed:
ER図、READMEの修正（２回目）

kei-sasaki-rgb-cell commented:
修正しました。

![](https://i.gyazo.com/0b5ca93c4623e578c90b86e9e25ba30a.png)

yuri240 reviewed (commented):
修正とレビュー依頼ありがとうございます。

問題なく設計できていることが確認できましたので、
LGTMとさせていただきます！

こちらのブランチは忘れずにMergeを行っていただき、
ローカルのMasterにPullしておきましょう。

データベース設計の実装、本当にお疲れ様でした。
引き続き、応援しております。
